### PR TITLE
#196 - Kafka fails to start when Zookeeper has SASL and SSL enabled

### PR DESCRIPTION
**Pull request by Pakkiyarasu**
Description of the change

Introduced KAFKA_ZOOKEEPER_TLS_CONFIG_PROPERTIES_FILE envrionment variable.

Modified code to support if Zookeeper is secured with SASL and SSL. I have added --zk-tls-config-file in kafka-configs.sh command if zookeeper protocol is SASL_SSL.

Created new file named zk-tls-config.properties while zookeeper_get_tls_config() method.

Benefits

we can configure Kafka if zookeeper has enbaled SASL authentication and SSL encryption.

Possible drawbacks

Applicable issues

Additional information

### CODE REVIEW

File: 2/debian-10/rootfs/opt/bitnami/scripts/kafka-env.sh [hunk 1]
```diff
@@ -44,7 +44,7 @@ kafka_env_vars=(
     KAFKA_ZOOKEEPER_TLS_TRUSTSTORE_PASSWORD
     KAFKA_ZOOKEEPER_TLS_TRUSTSTORE_FILE
     KAFKA_ZOOKEEPER_TLS_VERIFY_HOSTNAME
-    KAFKA_ZOOKEEPER_TLS_TYPE
+    KAFKA_ZOOKEEPER_TLS_TYPE
```
Comment (alemorcuq) [inline 2/debian-10/rootfs/opt/bitnami/scripts/kafka-env.sh:47]:
```suggestion
    KAFKA_ZOOKEEPER_TLS_TYPE
```

File: 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh [hunk 2]
```diff
@@ -445,7 +445,7 @@ EOF
 #   None
 #########################
 kafka_create_sasl_scram_zookeeper_users() {
-    export KAFKA_OPTS="-Djava.security.auth.login.config=${KAFKA_CONF_DIR}/kafka_jaas.conf"
+    export KAFKA_OPTS="$KAFKA_OPTS -Djava.security.auth.login.config=${KAFKA_CONF_DIR}/kafka_jaas.conf"
```
Comment (alemorcuq) [inline 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh:448]: Why this?

File: 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh [hunk 3]
```diff
@@ -456,7 +456,12 @@ kafka_create_sasl_scram_zookeeper_users() {
     for ((i = 0; i < ${#users[@]}; i++)); do
         debug "Creating user ${users[i]} in zookeeper"
         # Ref: https://docs.confluent.io/current/kafka/authentication_sasl/authentication_sasl_scram.html#sasl-scram-overview
-        debug_execute kafka-configs.sh --zookeeper "$KAFKA_CFG_ZOOKEEPER_CONNECT" --alter --add-config "SCRAM-SHA-256=[iterations=8192,password=${passwords[i]}],SCRAM-SHA-512=[password=${passwords[i]}]" --entity-type users --entity-name "${users[i]}"
+        if [[ "${KAFKA_ZOOKEEPER_PROTOCOL:-}" =~ SSL ]]; then
+            ZOOKEEPER_SSL_CONFIG=$(zookeeper_get_tls_config)
```
Comment (alemorcuq) [inline 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh:460]: We are not using `ZOOKEEPER_SSL_CONFIG` anywhere, are we?

File: 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh [hunk 4]
```diff
@@ -599,12 +604,27 @@ zookeeper_get_tls_config() {
...
           -Dzookeeper.client.secure=true \
-          -Dzookeeper.ssl.keyStore.location=${keystore_location} \
-          -Dzookeeper.ssl.keyStore.password=${KAFKA_ZOOKEEPER_TLS_KEYSTORE_PASSWORD} \
-          -Dzookeeper.ssl.trustStore.location=${kafka_zk_truststore_location} \
-          -Dzookeeper.ssl.trustStore.password=${KAFKA_ZOOKEEPER_TLS_TRUSTSTORE_PASSWORD} \
+          -Dzookeeper.ssl.keystore.location=${keystore_location} \
+          -Dzookeeper.ssl.keystore.password=${KAFKA_ZOOKEEPER_TLS_KEYSTORE_PASSWORD} \
+          -Dzookeeper.ssl.truststore.location=${kafka_zk_truststore_location} \
+          -Dzookeeper.ssl.truststore.password=${KAFKA_ZOOKEEPER_TLS_TRUSTSTORE_PASSWORD} \
+          -Dzookeeper.ssl.quorum.hostnameVerification=${KAFKA_ZOOKEEPER_TLS_VERIFY_HOSTNAME} \
```
Comment (alemorcuq) [inline 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh:627]: Why are we adding this new property?

**Review by alemorcuq — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh (hunk 2 above)
Comment (Pakkiyarasu) [reply 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh:448]: I am passing -Dzookeeper.ssl.hostnameVerification=false -Dzookeeper.ssl.quorum.hostnameVerification=false in KAFKA_OPTS during deployment. But this is getting overriden here. That's I added this change.

File: 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh (hunk 3 above)
Comment (Pakkiyarasu) [reply 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh:460]: `Yes we are not. But in zookeeper_get_tls_config() method, I am filling zk-tls-config.properties file. That's why I called this method here. If we don't assign to a variable, SSL settings are getting exposed in the logs. Thats why I assigned to a variable. `

File: 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh (hunk 4 above)
Comment (Pakkiyarasu) [reply 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh:627]: Without this, it was throwing hostname and address mismatch error when kafka tries to connect zookeeper.

File: 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh (hunk 3 above)
Comment (juan131) [reply 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh:460]: In the [run.sh](https://github.com/bitnami/bitnami-docker-kafka/blob/master/2/debian-10/rootfs/opt/bitnami/scripts/kafka/run.sh#L23), we're appending the output returned by `zookeeper_get_tls_config` to `KAFKA_OPTS`:
```
if [[ "${KAFKA_ZOOKEEPER_PROTOCOL:-}" =~ SSL ]]; then
    ZOOKEEPER_SSL_CONFIG=$(zookeeper_get_tls_config)
    export KAFKA_OPTS="$KAFKA_OPTS $ZOOKEEPER_SSL_CONFIG"
fi
```
Couldn't we do the same here? This way you don't need to create any `zk-tls-config.properties` file.

**Review by juan131 — CHANGES_REQUESTED** [review]
Thanks for this great PR! Please check my concerns when you have a chance.

**Review by Pakkiyarasu — COMMENTED** [review]
Made changes based on @juan131 comments

File: 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh (hunk 3 above)
Comment (Pakkiyarasu) [reply 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh]: Thanks for your inputs. I made above changes. It works fine. Code is much cleaner now.

**Review by juan131 — APPROVED** [review]
Thanks for applying the suggestions! LGTM
@alemorcuq could you take a look to the new changes too?

**Review by alemorcuq — APPROVED** [review]
Very nice! Thanks!

**Pull request merged**